### DISCUSSION TITLE
Improve user experience (UX)

### DISCUSSION THREAD

**Discussion by Gelassen** [remix-run/remix#3975] [Proposals]
I feel lack of "solidity as language" abilities to write things I usually code in others languages: at this moment this are data structures, dynamic arrays, generics. 

1. Data structures. 
In solidity I know about arrays, structs and mapping. It was enough for toying with ethereum abilities, but became very limiting when I started write ethereum extension for my project. Many users experience similar issues and already started to write own data structures based on contract ([ref](https://ethereum.stackexchange.com/questions/13167/are-there-well-solved-and-simple-storage-patterns-for-solidity)). At the same time it brings concerns about quality and security of such tools which is top-priority for finance purposed project which are majority in ethereum tech stack. Could the dev team aggregate demands&existing solutions in this aspect and start lead the development? 

2. Dynamic arrays. 

Dynamic arrays seems possible to define like this
```
address[] _users;
_users.push(addr); 
```
but remix compiler warns for the case of passing it into another function:
```
        address[] storage matches;
        for (uint idx = 0; idx < _demandsByUsers[user].length; idx++) {
            string memory demand = _demandsByUsers[user][idx];
            if (_indexNfts[demand].length != 0) { _addToCollection(matches, _indexNfts[demand]); }
        }
```
`TypeError: This variable is of storage pointer type and can be accessed without prior assignment, which would lead to undefined behaviour.`

References from many places including docs somewhat misleading that create a picture solidity arrays are dynamic when they are not, 

3. Generics and templates. 
I had not found any references on templates or generics in the official documentation, but for example [here](https://github.com/ethereum/solidity/issues/869) I see it is possible to have generics in solidiity. At another hand remix linter&compiler this code samples marks with error. Does remix linter isn't correct here or this code samples doesn't relevant to current version of solidiity?

**Comment by sergiodxa** [comment]
This repository is for [Remix](https://remix.run/) the web framework, not [Remix IDE](https://github.com/ethereum/remix-ide).

**Reply by Gelassen** [reply]
Thanks for point out on this, you are right. May be it is even better to forward to solidity repo in general.

**Reply by Gelassen** [reply]
I do not see the way to close the thread from my side, but the thread should be closed.
